### ISSUE TITLE
couchbaseCheckpoint added to all index mappings in Elasticsearch cluster

### ISSUE THREAD

**Issue by Elachance** [couchbaselabs/elasticsearch-transport-couchbase#47]
I'm working with an Elasticsearch cluster that handles data from many different sources and am hoping to get it processing Couchbase documents as well. However, once transport-couchbase has been installed on the ES cluster, all indexes in the cluster have this section added to their mappings:

```json
"couchbaseCheckpoint" : {
  "_source" : {
    "includes" : [
      "meta.*",
      "doc.*"
    ]
  },
  "dynamic_templates" : [
    {
      "store_no_index" : {
        "mapping" : {
          "index" : "no",
          "include_in_all" : false,
          "store" : "no"
        },
        "match" : "*"
      }
    }
  ],
  "properties" : {
    "meta" : {
      "include_in_all" : false,
      "type" : "object"
    }
  }
}
```
I don't know if this is normal or expected behavior, but it's unwanted. I've changed the first line of the template from:

```json
"template" : "*",
```

to

```json
"template" : "couch-*",
```
 
in an attempt to apply the couchbaseCheckpoint part of the mapping only to the index I intend to use for Couchbase, but it has not stopped this from happening.

What is adding this to all the indexes and what can I do to prevent it from continuing?

**Comment by g-eorge** [comment]
I'm experiencing the same issue. Are you saying this could be a problem in the Elasticsearch Java client?

**Comment by mschoch** [comment]
The behavior observed is that all index mappings (not documents) end up getting a section for "couchbaseCheckpoint".  This is expected as the template we provide contains "template": "*" which means it applies to all indexes.

It is harmless, as unless you actually create documents for this type, nothing will happen.

However, if you don't even want this section in your mapping.  You should change "*" in the template to be some pattern that only matches your couchbase indexes.  For example, you could prefix all you couchbase indexes with "cb" and then use "cb*".

**Comment by g-eorge** [comment]
Thanks for your response. I've been through the same process as the reporter..

- I set a prefix on the couchbase template
- Ensured it was updated in elasticsearch
- Deleted all existing indices
- Created a new empty test index
- Created a new couchbase index
- Created and started a couchbase -> elasticsearch replication

I observe `couchbaseCheckpoint` being added to the test index mappings when the couchbase replication gets underway.

**Comment by mschoch** [comment]
Let me number those steps for clarity:

1. I set a prefix on the couchbase template
2. Ensured it was updated in elasticsearch
3. Deleted all existing indices
4. Created a new empty test index
5. Created a new couchbase index
6. Created and started a couchbase -> elasticsearch replication

Can you provide the output of:

curl http://locahost:9200/test/_mapping

At 3 different points.

1.  After step 4.
2.  After step 5.
3.  After step 6.

Thanks

**Comment by g-eorge** [comment]
Sure.

After step 4 I have

`{}`

After step 5 I have

`{}`

After step 6 I have

```
{
   "test": {
      "mappings": {
         "couchbaseCheckpoint": {
            "properties": {
               "doc": {
                  "properties": {
                     "uuid": {
                        "type": "string"
                     }
                  }
               }
            }
         }
      }
   }
}
```

I have used the default template with "template": "cb*" and my couchbase index is called "cb_test".

**Comment by mschoch** [comment]
In that case it is a bug...

I suspect what is happening is that when Couchbase Server is invoking the discovery APIs like /pools, /pools/default, etc that one of these is inadvertently creating the bucket UUID.

I'm not exactly sure how we would fix that.  We need to list all of the Elasticsearch indexes as possible "destination buckets" for replication.  And the API requires we also provide the UUID for each of those buckets.  And we have to store the bucket UUIDs somewhere (currently we store it in the bucket) which is why we're seeing mapping change at that late stage.

So yes, this is indeed a bug.  One that the new maintainer will have to take a look at.

**Comment by prabitharp** [comment]
## ElasticSearch couchbase replication Issue


 I'm Trying to solve the issue while replicating my data from couch base server to elastic search.

*Error while replicating : Object mapping [place] trying to serialize a value with no field associated with it, current value [placeName]*

I have declared place as array of string in Person document.and place is re declared in another document called place document.


```Person : {"name":"..","place":["ppopo"]}```
 ```Place : {"place":{"name":"...","otherfields":"..."}}```

For see detailed issue please see this link : 

Couchbase ElasticSearch Replication Failure.
[How to solve the indexing error in couchbase elasticsearch replication](http://stackoverflow.com/questions/28361349/how-to-solve-the-indexing-error-in-couchbase-elasticsearch-replication)

**Comment by pumpkiny9120** [comment]
Hi guys, before I've seen the checkpoint docs in some indices and they were harmless. But recently we started using a daily rolling index with an alias pointing to all of them, and here's the error I saw in the log:
[2015-02-24 16:43:45,764][WARN ][org.eclipse.jetty.servlet.ServletHandler] /pools/default/buckets

org.elasticsearch.ElasticsearchIllegalArgumentException: Alias [phone_dist] has more than one index associated with it [[phone_dist-2015.02.22, phone_dist-2015.02.24, phone_dist-2015.02.13, phone_dist-2015.02.14, phone_dist-2015.02.18, phone_dist-2015.02.17, phone_dist-2015.02.12, phone_dist-2015.02.11, phone_dist-2015.02.23, phone_dist-2015.02.15, phone_dist-2015.02.16, phone_dist-2015.02.21, phone_dist-2015.02.20, phone_dist-2015.02.19, phone_dist-2015.02.25]], can't execute a single index op

Looks like this happened when the plugin is trying to insert a couchbaseCheckpoint document to the alias phone_dist, which is NOT part of the replication ("template": "phone") and it's been throwing errors all day since we had the alias there.
So far I'm not sure if the replication is still working okay. I'll keep an eye on that.
In the mean while, does anyone have a temporary workaround to this so the plugin can initialize normally?

**Comment by Branor** [comment]
@Elachance Hi, I'm the new dev for the plugin. As Marty said, the behavior you're seeing is indeed due to the fact that the plugin generates a UUID for every index (faux-bucket, as far as Couchbase is concerned) and stores it in the index itself. I'll add a config option to specify which buckets the plugin should use and which to ignore, so you can avoid having checkpoint docs stored in every index. This isn't a critical issue, so it won't make it into the upcoming release, but I'll make sure to add it to the next one after that.

**Comment by mukulsud** [comment]
@Branor this needs to be reopened since using the installation documentation that cb provides results in a Kibana breaking also.

The CB template is breaking this functionality in Kibana and causing this issue: https://github.com/elastic/kibana/issues/3331

After resetting the template, and reindexing our data I was able to get Kibana to set a default index. If you leave "template" : "*" (as recommended by the cb docs), drop all index patterns in Kibana you cannot create a new default index pattern.

**Comment by PhaedrusTheGreek** [comment]
+1 for a fix on this issue.  

By default, the plugin does not work with the Elasticsearch Marvel plugin, because .marvel indices have a strict mapping definition.

A nice default would be for indices beginning with a "." to be ignored.

**Comment by rostiar** [comment]
+1

**Comment by Branor** [comment]
Hi folks, I've added a potential fix for this issue with two new config options that let you specify which indexes the plugin is allowed to touch. Please take a look here: https://github.com/couchbaselabs/elasticsearch-transport-couchbase/releases/tag/2.1.2

Note that you also need to change the template that comes with the plugin if you want the mappings to only be added to specific indexes. Unfortunately, it's not possible to specify which index names it shouldn't apply to in the template, it only uses simple wildcards. So in addition to specifying a white-list of indexes the plugin can touch in elasticsearch.yml, you can have all Couchbase-related indexes have a common prefix (like cb*) and change the template to only apply to those.

**Comment by PhaedrusTheGreek** [comment]
Tested successfully for me!  Thank you!

@rostiar may have test feedback also...

**Comment by Branor** [comment]
Added the ability to ignore or whitelist both indexes and aliases. The version is released here: https://github.com/couchbaselabs/elasticsearch-transport-couchbase/releases/tag/2.1.2-GA
